cli/dstack/utils/interpolator.py

```diff
@@ -0,0 +1,68 @@
+import string
+from typing import Dict, Iterable, List, Optional, Tuple, Union
+
+
+class Pattern:
+    opening = "${{"
+    closing = "}}"
+
+
+class Name:
+    first_char = set(string.ascii_letters + "_")
+    char = first_char | set(string.digits + ".")
+
+
+class VariablesInterpolator:
+    def __init__(
+        self, namespaces: Dict[str, Dict[str, str]], *, skip: Optional[Iterable[str]] = None
+    ):
+        self.skip = set(skip) if skip is not None else set()
+        self.variables = {f"{ns}.{k}": v for ns in namespaces for k, v in namespaces[ns].items()}
+
+    def interpolate(
+        self, s: str, return_missing: bool = False
+    ) -> Union[str, Tuple[str, List[str]]]:
+        tokens = []
+        missing = []
+        start = 0
+        while start < len(s):
+            dollar = s.find("$", start)
+            if dollar == -1 or dollar == len(s) - 1:
+                tokens.append(s[start:])
+                break
+            if s[dollar + 1] == "$":  # escaped $$
+                tokens.append("$")
+                start += 2
+                continue
```

Comment:
Seems like the code doesn't handle escaped $$ in the middle of the string. E.g. `self.assertEqual("echo ${{ENV}}", get_interpolator().interpolate("echo $${{ENV}}"))` fails.